// twilio-mcs-client/lib/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const MCS_HOST = 'https://mcs.us1.twilio.com';
const MINIMUM_RETRY_DELAY = 1000;
const MAXIMUM_RETRY_DELAY = 4000;
const MAXIMUM_ATTEMPTS_COUNT = 3;
const RETRY_WHEN_THROTTLED = true;
class Configuration {
    constructor(token, baseUrl, options) {
        this.options = options.MCS || options || {};
        this.token = token;
        this.baseUrl =
            ((typeof this.options.realm === 'undefined') ? MCS_HOST : `https://mcs.${this.options.realm}.twilio.com`)
                + baseUrl;
    }
    static get backoffConfigDefault() {
        return {
            min: MINIMUM_RETRY_DELAY,
            max: MAXIMUM_RETRY_DELAY,
            maxAttemptsCount: MAXIMUM_ATTEMPTS_COUNT
        };
    }
    get backoffConfigOverride() {
        return this.options.backoffConfigOverride;
    }
    static get retryWhenThrottledDefault() { return RETRY_WHEN_THROTTLED; }
    get retryWhenThrottledOverride() { return this.options.retryWhenThrottledOverride; }
    updateToken(token) {
        this.token = token;
    }
}
exports.Configuration = Configuration;
